File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;
class Solution {
    public int solution(int n, int[][] edge) {
        ArrayList<Integer>[] graph = new ArrayList[n+1];
        for (int i = 0; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }   
        for(int[] e : edge){
            graph[e[0]].add(e[1]);
            graph[e[1]].add(e[0]);
        }
        
        int[] visited = new int[n+1];
        int max = 0;
        visited[1] = 1;
        
        Queue<int[]> q = new LinkedList<>();
        for(int g : graph[1]){
            q.add(new int[]{g, 1});
            visited[g] = 1;
        }
        
        while(!q.isEmpty()){
            int[] now = q.poll();
            for(int g : graph[now[0]]){
                if(visited[g] == 0){
                    q.add(new int[]{g, now[1]+1});
                    visited[g] = now[1]+1;
                    max = Math.max(max, now[1]+1);
                }
            }
        }
        
        int ans = 0;
        for(int v : visited){
            if(max == v){
                ans++;
            }
        }
        return ans;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    let graph = [];
    for(let i = 0; i <= n; i++){
        graph.push([]);
    }
    
    for(let i = 0; i < edge.length; i++){
        let a = edge[i][0];
        let b = edge[i][1];
        graph[a].push(b);
        graph[b].push(a);
    }
    
    let visited = Array(n+1).fill(0);
    visited[1] = 1;
    
    let q = [];
    q.push([1, 1]);
    
    let max = 0;
    
    while(q.length > 0){
        let now = q.shift();
        
        for(const num of graph[now[0]]){
            if(visited[num] === 0){
                q.push([num, now[1]+1]);
                max = Math.max(now[1]+1, max);
                visited[num] = now[1]+1;
            }
        }
    }
    
    let answer = 0;
    for(const v of visited){
        if(max === v){
            answer++;
        }
    }
    return answer;
}